### PR TITLE
feat(tests, cli): CLI tool for executing blocks against the live chain state

### PR DESCRIPTION
**Pull request by ekovalev**
A custom tool to replay a block against the live chain state (at some height, or the latest finalized block).
Essentially, a simplification of the `gear try-runtime` command that hides away the complexity of the latter and also allows to:
- not enable the `try-runtime` feature during node/runtime build;
- use the native runtime to apply a block, that is run extrinsics in a debugger, if needed.

See the `README.md` for more details.

### CODE REVIEW

**Comment by NikVolf** [comment]
seems useful

@ukint-vs can you try this and review?

**Comment by ekovalev** [comment]
> Im getting `ERROR runtime: panicked at 'Bad input data provided to execute_block: Codec error', runtime/gear/src/lib.rs:651:1` every time when I try to execute block with some `pallet_gear` tx in local dev network.

The reason for errors may have been found: since we are using the `Core` `runtime_api` (specifically, `Executive::execute_block(block)`), the final checks are also run by this runtime api implementation (the `executive` pallet). In case we don't have extrinsics in a block these checks go through, but if we run extrinsics, the state resulting root doesn't match the one downloaded from the live chain.

The only option I can think of now is to enable the `try-runtime` feature and use the `try_execute_block` runtime api call instead (which allows to skip the state root checks). This would mean we will have to only use the native runtime (because the downloaded one from the live chain will obviously not have the `try-runtime` feature enabled).

**Comment by breathx** [comment]
> The reason for errors may have been found: since we are using the `Core` `runtime_api` (specifically, `Executive::execute_block(block)`), the final checks are also run by this runtime api implementation (the `executive` pallet). In case we don't have extrinsics in a block these checks go through, but if we run extrinsics, the resulting state root doesn't match the one downloaded from the live chain.
> 
> The only option I can think of now is to enable the `try-runtime` feature and use the `try_execute_block` runtime api call instead (which allows to skip the state root checks). This would mean we will have to only use the native runtime (because the downloaded one from the live chain will obviously not have the `try-runtime` feature enabled).

Special RPC node with the feature wouldn't be the solution?

File: utils/remote-ext-tests/replay-block/src/main.rs
Comment (ekovalev) [reply utils/remote-ext-tests/replay-block/src/main.rs:88]: Done.

File: utils/remote-ext-tests/replay-block/README.md
Comment (breathx) [inline utils/remote-ext-tests/replay-block/README.md:50]: can't we use block number here?

File: utils/remote-ext-tests/replay-block/src/main.rs
Comment (breathx) [inline utils/remote-ext-tests/replay-block/src/main.rs:94]:
```suggestion
    #[arg(short = 'l', long, value_name = "NODE_LOG", num_args = 0..)]
```

Comment (breathx) [reply utils/remote-ext-tests/replay-block/src/main.rs:94]: Same as RUST_LOG

File: utils/remote-ext-tests/replay-block/src/main.rs
Comment (breathx) [inline utils/remote-ext-tests/replay-block/src/main.rs:145]: I see that number is possible so state it in the docs please

File: utils/remote-ext-tests/replay-block/src/main.rs
Comment (breathx) [inline utils/remote-ext-tests/replay-block/src/main.rs:40]: too complicated name for target for me, it's not as comfort to use as could be with for example `gear-replay`
And I'd then consider crate renaming into something similar and not such general just because it supports concrete functionality of us as separate and the only one project

File: utils/remote-ext-tests/replay-block/src/parse.rs
Comment (breathx) [inline utils/remote-ext-tests/replay-block/src/parse.rs:39]: Open an issue maybe and ref it here

File: utils/remote-ext-tests/replay-block/src/parse.rs
Comment (breathx) [inline utils/remote-ext-tests/replay-block/src/parse.rs:54]:
```suggestion
            format!("Expected block hash or number, found illegal hex character at position {e}")
```
Everywhere where possible please

File: utils/remote-ext-tests/replay-block/src/util.rs
Comment (breathx) [inline utils/remote-ext-tests/replay-block/src/util.rs:87]: deadcode?

**Review by breathx — APPROVED** [review]
Awesome

File: utils/remote-ext-tests/replay-block/src/parse.rs
Comment (ekovalev) [reply utils/remote-ext-tests/replay-block/src/parse.rs:39]: It's a leftover from the original `try-runtime`. I better remove the comment completely.

File: utils/remote-ext-tests/replay-block/README.md
Comment (ekovalev) [reply utils/remote-ext-tests/replay-block/README.md:50]: this refers to the original `try-runtime` CLI. In our tool we do support both the hash and block number

File: utils/remote-ext-tests/replay-block/src/main.rs
Comment (ekovalev) [reply utils/remote-ext-tests/replay-block/src/main.rs:145]: The `README.md` has the full description. It'a just the first part refers to the in-built `try-runtime` node subcommand (and how it can be used for the similar purpose).

File: utils/remote-ext-tests/replay-block/src/util.rs
Comment (ekovalev) [reply utils/remote-ext-tests/replay-block/src/util.rs:87]: It's not used anymore, kept it in case we need something like this in the future (parse a string literal into a block hash). However, it's quite trivial anyway - I think it can be safely removed for now

File: utils/remote-ext-tests/replay-block/src/main.rs
Comment (ekovalev) [reply utils/remote-ext-tests/replay-block/src/main.rs:94]: Why `num_args = 0..`? Do we want to allow a dangling `-l` flag with no target?

File: utils/remote-ext-tests/replay-block/src/parse.rs
Comment (breathx) [reply utils/remote-ext-tests/replay-block/src/parse.rs:39]: > It's a leftover from the original `try-runtime`. I better remove this comment completely.

lol

File: utils/remote-ext-tests/replay-block/src/main.rs
Comment (breathx) [reply utils/remote-ext-tests/replay-block/src/main.rs:94]: > Why `num_args = 0..`? Do we want to allow a dangling `-l` flag with no target?

for example if we will throw some env var like `-l LOG_ENV` so why not

**Comment by breathx** [comment]
Waiting for last minor fixes @ekovalev and review from @ukint-vs

**Review by ukint-vs — APPROVED** [review]

**Comment by breathx** [comment]
@ekovalev merge if ready

**Pull request merged**
